src/contrib/mlir/core/pass/ng_dialect_builder.cpp

```diff
@@ -0,0 +1,541 @@
+#include "ngraph/graph_util.hpp"
+#include "ngraph/node.hpp"
+#include "ngraph/op/add.hpp"
+#include "ngraph/op/argmax.hpp"
+#include "ngraph/op/argmin.hpp"
+#include "ngraph/op/concat.hpp"
+#include "ngraph/op/convolution.hpp"
+#include "ngraph/op/divide.hpp"
+#include "ngraph/op/dot.hpp"
+#include "ngraph/op/experimental/compiled_kernel.hpp"
+#include "ngraph/op/gather.hpp"
+#include "ngraph/op/greater.hpp"
+#include "ngraph/op/less.hpp"
+#include "ngraph/op/maximum.hpp"
+#include "ngraph/op/minimum.hpp"
+#include "ngraph/op/multiply.hpp"
+#include "ngraph/op/negative.hpp"
+#include "ngraph/op/relu.hpp"
+#include "ngraph/op/subtract.hpp"
+#include "ngraph/op/util/index_reduction.hpp"
+#include "ngraph/type/element_type.hpp"
+
+#include "contrib/mlir/utils.hpp"
+
+#include <llvm/ADT/STLExtras.h>
+#include <llvm/Analysis/TargetTransformInfo.h>
+#include <llvm/ExecutionEngine/Orc/JITTargetMachineBuilder.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Support/ErrorOr.h>
+#include <llvm/Support/MemoryBuffer.h>
+#include <llvm/Support/SourceMgr.h>
+#include <llvm/Support/TargetSelect.h>
+#include <llvm/Target/TargetMachine.h>
+#include <mlir/Conversion/LoopToStandard/ConvertLoopToStandard.h>
+#include <mlir/Conversion/StandardToLLVM/ConvertStandardToLLVM.h>
+#include <mlir/Conversion/StandardToLLVM/ConvertStandardToLLVMPass.h>
+#include <mlir/Dialect/LLVMIR/LLVMDialect.h>
+#include <mlir/ExecutionEngine/ExecutionEngine.h>
+#include <mlir/ExecutionEngine/MemRefUtils.h>
+#include <mlir/ExecutionEngine/OptUtils.h>
+#include <mlir/Pass/PassManager.h>
+#include <mlir/Target/LLVMIR.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <mlir/Transforms/Passes.h>
+
+#include <memory>
+#include <mutex>
+
+// Defines a new LLVM debug type for this file to be used by LLVM_DEBUG macro.
+#define DEBUG_TYPE "mlir-compiler"
+
+using llvm::SmallVector;
+using llvm::StringRef;
+using llvm::ArrayRef;
+
+using namespace ngraph;
+using namespace ngraph::runtime::ngmlir;
+
+#define COMPILE_OP_DECL(op_name)                                                                   \
+    createOp<op_name>(NgDialectConversionPass & NgDialectObj, const ngraph::Node* ngNode)
+
+namespace
+{
+    /// NgDialectConversionPass is an MLIR ModulePass Given an nGraph sub-graph, represented as
+    /// CompiledKernel node, it
+    /// translates the graph down to nGraph dialect
+
+    class NgDialectConversionPass : public mlir::ModulePass<NgDialectConversionPass>
+    {
+    public:
+        using TensorList = std::vector<descriptor::Tensor*>;
+        using TypeList = llvm::SmallVector<mlir::Type, 4>;
+
+        NgDialectConversionPass(const ngraph::op::CompiledKernel* compiled_kernel,
+                                mlir::MLIRContext* context,
+                                mlir::OpBuilder* builder,
+                                MLIRCompiler* compiler)
+            : m_compiledKernel(compiled_kernel)
+            , m_context(context)
+            , m_builder(builder)
```

Comment:
See comment below about builder. Since we are passing a context here, we should be able to create a local builder object here. We can set builder's insertion point in runOnFunction, once we have the module to be processed.